Update host card instance preview when edited in outside of app

habdelra opened the pull request.

habdelra committed:
Add index invalidation to SSE messages

habdelra committed:
live update card instances in code route

habdelra committed:
Merge remote-tracking branch 'origin/main' into cs-5267-bug-when-a-card-source-is-edited-in-vs-code-the-indexer

habdelra committed:
add live update for code mode preview card

habdelra commented:
One thing I noticed is that the code route preview is more stable (there is less glimmer invalidation). I have a hunch that the resource might need a stale value while the load is in progress. (I saw this in the cards-grid module too when I was making that work for live loading). I'll look into that a bit more...

jurgenwerk commented on packages/runtime-common/realm.ts:
```diff
@@ -228,21 +236,84 @@ export class Realm {
+      onInvalidation: (invalidatedURLs: URL[]) => {
+        this.sendUpdateMessages({
+          type: 'update',
+          data: { 'index-invalidation': JSON.stringify(invalidatedURLs) },
```
Would it make sense to put "index-invalidation" value in an enum of event names?

jurgenwerk reviewed (commented):
I have tested this locally and I think I found a bug. If I go edit an instance (for example I change `firstName` in `packages/drafts-realm/Friend/1.json`), the preview will reload (I see it flicker), but the change will not be reflected (name stays the same). But if I save for the 2nd time, the change will appear.

tintinthong commented:
> I have tested this locally and I think I found a bug. If I go edit an instance (for example I change `firstName` in `packages/drafts-realm/Friend/1.json`), the preview will reload (I see it flicker), but the change will not be reflected (name stays the same). But if I save for the 2nd time, the change will appear.

I concur with this observation. I tested by opening the monaco file in code route in another tab. The first change isnt picked up but the second one gets picked up. This is despite viewing the 2 incremental indexing events when listening in `curl -H "Accept: text/event-stream" http://localhost:4201/drafts/_message`

tintinthong commented:
Do I expect this to work for link fields? Like if I edit Person/1.json will Booking.Hosts renders be reflected when looking at isolated view. From what I see the updated data of the links don't render

habdelra commented:
> Do I expect this to work for link fields? Like if I edit Person/1.json will Booking.Hosts renders be reflected when looking at isolated view. From what I see the updated data of the links don't render

not yet. as I mentioned in the last boxel meeting card instances don't know their dependencies yet. as soon as we implement that, then yes this should work automatically.

habdelra commented:
> > I have tested this locally and I think I found a bug. If I go edit an instance (for example I change `firstName` in `packages/drafts-realm/Friend/1.json`), the preview will reload (I see it flicker), but the change will not be reflected (name stays the same). But if I save for the 2nd time, the change will appear.
> 
> I concur with this observation. I tested by opening the monaco file in code route in another tab. The first change isnt picked up but the second one gets picked up. This is despite viewing the 2 incremental indexing events when listening in `curl -H "Accept: text/event-stream" http://localhost:4201/drafts/_message`

i think i know what this might be--i had to eliminate an echo that would cause the events of our own writes to re-trigger an index. perhaps that echo cancellation is getting in the way

habdelra committed:
Merge remote-tracking branch 'origin/main' into cs-5267-bug-when-a-card-source-is-edited-in-vs-code-the-indexer

habdelra committed:
disable echo cancellation for debugging

habdelra commented:
@tintinthong @jurgenwerk I'm unable to reproduce the issue you guys found, every change that I make (whether in a separate tab via code route editor or VS code) is reflected in the instance preview in code mode and the code route. So i think i'll need to fix this blind. I disabled the echo cacncellation, can you pull the latest and see if you still see the problem? perhaps you can view the `http://localhost:4201/drafts/_message` request in your XHR logs and show me what the event stream looks like?

lukemelia commented:
I noticed that in Code Mode, if the preview of the card is showing the Embedded or Edit view, a remote edit of the card results in the view reverting back to Isolated. Not sure if this should be considered a bug for this PR.

habdelra commented:
@lukemelia did you also notice the issues that Justin and Matic ran into?

lukemelia commented on packages/host/app/components/operator-mode/code-mode.gts:
```diff
@@ -28,13 +30,36 @@ interface Signature {
+    registerDestructor(this, () => {
+      if (this.subscription) {
+        this.subscription.unsubscribe();
+      }
```
```suggestion
      this.subscription?.unsubscribe();
```

lukemelia commented on packages/host/app/components/editor/go.gts:
```diff
@@ -139,11 +144,34 @@ export default class Go extends Component<Signature> {
+    registerDestructor(this, () => {
+      if (this.subscription) {
+        this.subscription.unsubscribe();
+      }
```
```suggestion
      this.subscription?.unsubscribe();
```

lukemelia commented on packages/runtime-common/realm.ts:
```diff
@@ -926,9 +997,24 @@ export class Realm {
-      );
+      await this.#adapter.subscribe((data: Record<string, any>) => {
+        let tracked = this.getTrackedWrite(data);
+        // TODO debugging....
```
What's the status of this comment?

lukemelia commented:
> @lukemelia did you also notice the issues that Justin and Matic ran into?

No I couldn't reproduce that.

lukemelia commented:
> @lukemelia did you also notice the issues that Justin and Matic ran into?

No I couldn't reproduce that.

habdelra committed:
Typed our SSE events and re-introduced echo elimination

habdelra commented:
@lukemelia I reintroduced the echo elimination to this PR. do you wanna pull and try one more time and see if you can see the issues that Matic and Justin ran into? thanks!

habdelra commented:
for some reason I can't reply to comments in the code:

@lukemelia the status of the "debugging" comment was for Justin and Matic to get back to me on what they were seeing (this is the removal of the echo handling code). I just reverted it since it sounds like you don't see this either. I'm not too sure what is going on...

@jurgenwerk i refactored to type all the SSE events and their data, so it should be really clear what event data is allowed and what is not, and the types can guide us in development

habdelra committed:
updated based on review feedback

lukemelia commented:
@habdelra still works for me, although there is a debugger committed at the moment

tintinthong commented on packages/host/app/components/operator-mode/code-mode.gts:
```diff
@@ -28,13 +30,40 @@ interface Signature {
+          if (type !== 'index') {
+            return;
+          }
+          debugger;
```
debugger statement

habdelra replied:
oops!

tintinthong reviewed (approved):
@habdelra I pulled recents and did a fresh rebuild and I was unable to recreate the issue previously. Apologies if this was misleading. There is a remaining debugger statement

habdelra committed:
remove debugger
